digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/client/vvh/KgvOperation.scala

```diff
@@ -0,0 +1,582 @@
+          0
+      }
+    }
+    attributesMap.filterKeys { x =>
+      Set(
+        "roadnumber",
+        "roadpartnumber",
+        "municipalitycode",
+        "surfacetype",
+        "kmtkid",
+        "roadclass",
+        "xyaccuracy",
+        "zaccuracy",
+        "surfacerelation",
+        "lifecyclestatus",
+        "roadnamefin",
+        "roadnameswe",
+        "addressfromright",
+        "addresstoright",
+        "addressfromleft",
+        "addresstoleft",
+        "geometryflip",
+        "starttime",
+        "versionstarttime",
+        "sourcemodificationtime"
+      ).contains(x)
+    }.filter { case (_, value) =>
+      value != null
+    }
+  }
+
+  /**
+   * Extract double value from data. Used for change info start and end measures.
+   */
+  private def anyToDouble(value: Any): Option[Double] = {
+    value match {
+      case null => None
+      case _ => {
+        val doubleValue = Try(value.toString.toDouble).getOrElse(throw new NumberFormatException(s"Failed to convert value: ${value.toString}") )
+        Some(doubleValue)
+      }
+    }
+  }
+
+  def toBigInt(value: Int): BigInt = {
+    Try(BigInt(value)).getOrElse(throw new NumberFormatException(s"Failed to convert value: ${value.toString}"))
+  }
+
+  private def extractModifiedAt(attributes: Map[String, Any]): Option[DateTime] = {
+    val validFromDate = Option(new DateTime(attributes("sourcemodificationtime").asInstanceOf[String]).getMillis)
+    var lastEditedDate : Option[Long] = Option(0)
+    if(attributes.contains("versionstarttime")){
+      lastEditedDate = Option(new DateTime(attributes("versionstarttime").asInstanceOf[String]).getMillis)
+    }
+
+    lastEditedDate.orElse(validFromDate).map(modifiedTime => new DateTime(modifiedTime))
+  }
+
+  def extractFeature(feature: Feature, path: List[List[Double]], linkGeomSource: LinkGeomSource): RoadLinkFetched = {
+    val attributes = feature.properties
+
+    val validFromDate = Option(BigInteger.valueOf(new DateTime(attributes("sourcemodificationtime").asInstanceOf[String]).getMillis))
+    val lastEditedDate = Option(BigInteger.valueOf(new DateTime(attributes("versionstarttime").asInstanceOf[String]).getMillis))
+    val startTime = Option(BigInteger.valueOf(new DateTime(attributes("starttime").asInstanceOf[String]).getMillis))
+
+    val linkGeometry: Seq[Point] = path.map(point => {
+      Point(anyToDouble(point(0)).get, anyToDouble(point(1)).get, anyToDouble(point(2)).get)
+    })
+    val linkGeometryForApi = Map("points" -> path.map(point => Map("x" -> anyToDouble(point(0)).get, "y" -> anyToDouble(point(1)).get, "z" -> anyToDouble(point(2)).get, "m" -> anyToDouble(point(3)).get)))
+    val linkGeometryWKTForApi = Map("geometryWKT" -> (s"LINESTRING ZM (${path.map(point => anyToDouble(point(0)).get + " " + anyToDouble(point(1)).get + " " + anyToDouble(point(2)).get + " " + anyToDouble(point(3)).get).mkString(", ")})"))
```

Comment:
kannataa tehdä tikettti selvittää tarvitaanko näitä jossakin. tulee viemään vähän muistia kanniskella ympäri ohjelmistoa

Reply:
etenkin kun UUID tulee myös viemään muistia